src/psyclone/tests/psyir/transformations/loop_fusion_test.py

```diff
@@ -0,0 +1,585 @@
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              end subroutine sub'''
+    fuse_loops(parser, code)
+
+
+# ----------------------------------------------------------------------------
+def test_fuse_dimension_change(parser):
+    '''Test that inconsistent use of dimemsions are detected, e.g.:
+    loop1:  a(i,j)
+    loop2:  a(j,i)
+    when at least one operation is a write
+    '''
+
+    # The first example can be merged, since 't' is read-only,
+    # so it doesn't matter that it is accessed differently
+    code = '''subroutine sub()
+              integer :: ji, jj, n
+              integer, dimension(10,10) :: s, t
+              do jj=1, n+1
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              do jj=1, n+1
+                 do ji=1, 10
+                    s(ji, jj)=s(ji, jj) + t(jj, jj) + t(ji, ji)
+                 enddo
+              enddo
+              end subroutine sub'''
+
+    fuse_loops(parser, code)
+
+    # This cannot be fused, since 's' is written in the
+    # first iteration and read in the second.
+    code = '''subroutine sub()
+              integer :: ji, jj, n
+              integer, dimension(10,10) :: s, t, u
+              do jj=1, n+1
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              do jj=1, n+1
+                 do ji=1, 10
+                    u(ji, jj)=s(jj, ji)+1
+                 enddo
+              enddo
+              end subroutine sub'''
+
+    with pytest.raises(TransformationError) as err:
+        fuse_loops(parser, code)
+    assert "Variable 's' is using loop variable jj in index 1 and 0." \
+        in str(err.value)
+
+    # This cannot be fused, since 's' is read in the
+    # first iteration and written in the second with
+    # different indices.
+    code = '''subroutine sub()
+              integer :: ji, jj, n
+              integer, dimension(10,10) :: s, t, u
+              do jj=1, n+1
+                 do ji=1, 10
+                    u(ji, jj)=s(jj, ji)+1
+                 enddo
+              enddo
+              do jj=1, n+1
+                 do ji=1, 10
+                    s(ji, jj)=t(ji, jj)+1
+                 enddo
+              enddo
+              end subroutine sub'''
+
+    with pytest.raises(TransformationError) as err:
+        fuse_loops(parser, code)
+    assert "Variable 's' is using loop variable jj in index 0 and 1." \
+        in str(err.value)
+
+
+# ----------------------------------------------------------------------------
+def test_fuse_independent_array(parser):
+    '''Test that using arrays which are not dependent on the loop variable
+    are handled correctly. Example:
+    do j  ... a(1) = b(j) * c(j)
+    do j ...  d(j) = a(1)
+    '''
+
+    # The first example can be merged, since 't' is read-only,
```

Comment:
It's 's' that's independent of loop variable, oh, and it's written to in the first loop :-)

Reply:
Done.